plugins/logs/plugin.go

```diff
@@ -566,47 +602,45 @@ func (p *Plugin) loop() {
 
 	ctx, cancel := context.WithCancel(context.Background())
 
-	var retry int
-
 	for {
-		var err error
 
-		if p.config.Service != "" {
-			var uploaded bool
-			uploaded, err = p.oneShot(ctx)
+		var retry int
+		var waitC chan struct{}
 
-			if err != nil {
-				p.logger.Error("%v.", err)
-			} else if uploaded {
-				p.logger.Info("Logs uploaded successfully.")
-			} else {
-				p.logger.Debug("Log upload queue was empty.")
-			}
-		}
+		if *p.config.Reporting.Trigger == plugins.TriggerPeriodic && p.config.Service != "" {
+			err := p.doOneShot(ctx)
 
-		var delay time.Duration
+			var delay time.Duration
 
-		if err == nil {
-			min := float64(*p.config.Reporting.MinDelaySeconds)
-			max := float64(*p.config.Reporting.MaxDelaySeconds)
-			delay = time.Duration(((max - min) * rand.Float64()) + min)
-		} else {
-			delay = util.DefaultBackoff(float64(minRetryDelay), float64(*p.config.Reporting.MaxDelaySeconds), retry)
-		}
+			if err == nil {
+				min := float64(*p.config.Reporting.MinDelaySeconds)
+				max := float64(*p.config.Reporting.MaxDelaySeconds)
+				delay = time.Duration(((max - min) * rand.Float64()) + min)
+			} else {
+				delay = util.DefaultBackoff(float64(minRetryDelay), float64(*p.config.Reporting.MaxDelaySeconds), retry)
+			}
 
-		if p.config.Service != "" {
 			p.logger.Debug("Waiting %v before next upload/retry.", delay)
-		}
 
-		timer := time.NewTimer(delay)
+			waitC = make(chan struct{})
+			go func() {
+				select {
+				case <-time.After(delay):
+					if err != nil {
+						retry++
+					} else {
+						retry = 0
+					}
+					close(waitC)
+				case <-ctx.Done():
+					return
+				}
+			}()
+		}
 
 		select {
-		case <-timer.C:
-			if err != nil {
-				retry++
-			} else {
-				retry = 0
-			}
+		case <-waitC:
+			break
```

Comment:
[nit] We can drop this; select doesn't have fallthrough behaviour. Once a case is selected, only that one will be executed.
```suggestion
```

Reply:
Removed